src/nano_contracts/builder.ts

```diff
@@ -0,0 +1,328 @@
+  }
+
+  /**
+   * Build the nano contract transaction
+   *
+   * @memberof NanoContractTransactionBuilder
+   * @inner
+   */
+  async build(): Promise<NanoContract> {
+    if (this.blueprintId === null || this.method === null || this.caller === null) {
+      throw new Error('Must have blueprint id, method and caller.');
+    }
+
+    if (this.method !== 'initialize' && this.ncId === null) {
+      throw new Error(`Nano contract ID cannot be null for method ${this.method}`);
+    }
+
+    // Transform actions into inputs and outputs
+    let inputs: Input[] = [];
+    let outputs: Output[] = [];
+    let tokens: string[] = [];
+    if (this.actions) {
+      const tokenSet = new Set<string>();
+      for (const action of this.actions) {
+        // Get token list
+        if (action.token !== HATHOR_TOKEN_CONFIG.uid) {
+          tokenSet.add(action.token);
+        }
+      }
+      tokens = Array.from(tokenSet);
+      for (const action of this.actions) {
+        // Call action
+        if (action.type === NanoContractActionType.DEPOSIT) {
+          const ret = await this.executeDeposit(action, tokens);
+          inputs = concat(inputs, ret[0]);
+          outputs = concat(outputs, ret[1]);
+        } else if (action.type === NanoContractActionType.WITHDRAWAL) {
+          const output = this.executeWithdrawal(action, tokens);
+          outputs = concat(outputs, output);
+        } else {
+          throw new Error('Invalid type for nano contract action.');
+        }
+      }
+    }
+
+    // Serialize the method arguments
+    const serializedArgs: Buffer[] = [];
+    if (this.args) {
+      const serializer = new Serializer();
+      const blueprintInformation = await ncApi.getBlueprintInformation(this.blueprintId);
+      const methodArgs = get(blueprintInformation, `public_methods.${this.method}.args`, []) as MethodArgInfo[];
+      if (!methodArgs) {
+        throw new NanoContractTransactionError(`Blueprint does not have method ${this.method}.`);
+      }
+
+      if (this.args.length !== methodArgs.length) {
+        throw new NanoContractTransactionError(`Method needs ${methodArgs.length} parameters but data has ${this.args.length}.`);
+      }
+
+      for (const [index, arg] of methodArgs.entries()) {
+        let serialized: Buffer;
+        if (arg.type.startsWith('SignedData[')) {
+          const splittedValue = this.args[index].split(',');
+          if (splittedValue.length !== 3) {
+            throw new Error('Signed data requires 3 parameters.');
+          }
+          // First value must be a Buffer but comes as hex
+          splittedValue[0] = hexToBuffer(splittedValue[0]);
+          const tupleValues: [Buffer, any, string] = splittedValue;
+          if (tupleValues[2] === 'bytes') {
+            // If the result is expected as bytes, it will come here in the args as hex value
+            tupleValues[1] = hexToBuffer(tupleValues[1]);
+          }
+          serialized = serializer.fromSigned(...tupleValues);
```

Comment:
The `Serializer` class should know how to serialize the different types. If I understood correctly, you propose to move the code above inside `serializer.fromSigned`. If that's the case, I'm not sure I agree because the signed data standard with 3 parameters separated by comma was defined by the headless wallet to support this and other clients might implement it differently, that's why I added this validation in the builder.

Reply:
After we talked I understood what you meant. Move the validation to the serializer class method: https://github.com/HathorNetwork/hathor-wallet-lib/pull/604/commits/ae6f1a49a5dc218656a43f162ff4f3225c37f320